modules/localization/msf/local_integ/localization_gnss_process.cc

```diff
@@ -0,0 +1,437 @@
+      t_type != drivers::gnss::GnssType::GLO_SYS) {
+    return true;
+  }
+  unsigned int sat_prn = 0;
+  unsigned int week_num = 0;
+  double toe = -0.1;
+  if (t_type == drivers::gnss::GnssType::GLO_SYS) {
+    sat_prn = raw_eph.glonass_orbit().slot_prn();
+    week_num = raw_eph.glonass_orbit().week_num();
+    toe = raw_eph.glonass_orbit().toe();
+  } else {
+    sat_prn = raw_eph.keppler_orbit().sat_prn();
+    week_num = raw_eph.keppler_orbit().week_num();
+    toe = raw_eph.keppler_orbit().toe();
+  }
+  const EphKey temp(drivers::gnss::GnssType(t_type), sat_prn, week_num, toe);
+  if (map_gnss_eph_.find(temp) != map_gnss_eph_.end()) {
+    return true;
+  }
+  map_gnss_eph_.insert(
+      std::map<EphKey, drivers::gnss::GnssEphemeris>::
+      value_type(temp, raw_eph));
+  return false;
+}
+
+inline void LocalizationGnssProcess::LogPnt(const GnssPntResultMsg &rover_pnt,
+                                            double ratio) {
+  char print_infor[256] = {'\0'};
+  snprintf(print_infor, sizeof(print_infor),
+           "%6d%12.3f%4d%16.3f%16.3f%16.3f%4d%4.1f%6.1f%8.3f%8.3f%8.3f%8.3f%8."
+           "3f%8.3f\n",
+           rover_pnt.gnss_week(), rover_pnt.gnss_second_s(),
+           static_cast<int>(rover_pnt.pnt_type()),
+           rover_pnt.pos_x_m(), rover_pnt.pos_y_m(),
+           rover_pnt.pos_z_m(), rover_pnt.sovled_sat_num(), rover_pnt.pdop(),
+           ratio, rover_pnt.vel_x_m(), rover_pnt.vel_y_m(),
+           rover_pnt.vel_z_m(), rover_pnt.std_pos_x_m(),
+           rover_pnt.std_pos_y_m(), rover_pnt.std_pos_z_m());
+  LOG(INFO) << print_infor;
+}
+
+bool LocalizationGnssProcess::GnssPosition(
+    EpochObservationMsg *raw_rover_obs) {
+  gnss_state_ = LocalizationMeasureState::NOT_VALID;
+  if (raw_rover_obs->receiver_id() != 0) {
+    LOG(INFO) << "Wrong Rover Obs Data!";
+    return false;
+  }
+  // debug
+  int b_solved = gnss_solver_->solve(raw_rover_obs, &gnss_pnt_result_);
+  if (b_solved < 0) {
+    return false;
+  }
+  LogPnt(gnss_pnt_result_, gnss_solver_->get_ratio());
+  if (!sins_align_finish_) {
+    LOG(INFO) << "Sins-ekf has not converged or finished its aligment!";
+  }
+  if (gnss_pnt_result_.has_std_pos_x_m() &&
+      gnss_pnt_result_.has_std_pos_y_m() &&
+      gnss_pnt_result_.has_std_pos_z_m()) {
+    double sigma =
+        gnss_pnt_result_.std_pos_x_m() * gnss_pnt_result_.std_pos_x_m() +
+        gnss_pnt_result_.std_pos_y_m() * gnss_pnt_result_.std_pos_y_m() +
+        gnss_pnt_result_.std_pos_z_m() * gnss_pnt_result_.std_pos_z_m();
+    sigma = std::sqrt(fabs(sigma));
+    if (fabs(sigma) > 10.0) {
+      LOG(INFO) << "Position std exceeds the threshold 10.0!";
+      return false;
+    }
+  }
+
+  gnss_state_ = LocalizationMeasureState::OK;
+
+  return true;
+}
+
+bool LocalizationGnssProcess::Print(const char *print_infor,
+                                    const int print_times) {
+  for (int i = 0; i < print_times; ++i) {
+    printf("%s", print_infor);
```

Comment:
Please try avoid using unsafe functions such as printf.

Reply:
delete this function